### PR TITLE
compile_only_deps for cc_*

### PR DESCRIPTION
**Pull request by werkt**
These dependencies are used only to compile .cc files. Linking semantics
are the same as with 'deps'.

Change-Id: I565c39e767fe86b969331614445d27e5d64103d8

### CODE REVIEW

**Comment by hlopko** [comment]
CC @klimek

**Comment by oquenchil** [comment]
Hi George,

if we had include scanning (pruning headers) and layering check (making sure you are depending on the rules that have the headers you are using) then I believe the problem that you are solving with this PR would be solved.

We will have that in Bazel eventually but that's far away and I understand we have to find a solution meanwhile.

What is the main problem you are trying to solve with this PR? Is it that you pass too many headers to downstream actions that don't need it therefore making them slower and use more bandwidth in remote execution? Is it about header visibility? Downstream rules not giving compilation error when using headers they shouldn't use?

Also would this new attribute be used throughout your codebase or just in some key places? I ask this last question because we can do exactly the same thing your PR does already today in Bazel without any additional changes with the help of the C++ sandwich. If a lot of targets will be using the compile_only_deps attribute then the C++ sandwich idea would become really annoying. 

To summarize, the idea would be to have one cc_library with only the headers that you want to propagate. You would also have your cc_library with cc files depending on the previous headers library. 

Now you introduce a new rule written in Starlark called cc_linking_library that would depend on just a single cc_library (wrapping the cc files library above) and omit the compilation context when creating a new [CcInfo](https://docs.bazel.build/versions/master/skylark/lib/CcInfo.html) . 

The dependent rule can depend on this cc_linking_library and separately on any other headers that were meant to be propagated.

You can see how this would be extremely annoying if you need this in more than a couple of places. If that's the case, then I believe we have no choice but to add compile_only_deps, at least until include scanning and layering check are in Bazel.

**Comment by werkt** [comment]
> if we had include scanning (pruning headers) and layering check (making sure you are depending on the rules that have the headers you are using) then I believe the problem that you are solving with this PR would be solved.

Probably. I've been promised include scanning for 3 years, I don't believe it will be available soon, and I think that the non-cpp side effects of massive numbers of inputs flowing from low level dependencies that don't intend to present them as their interface is substantial, details below.
> 
> We will have that in Bazel eventually but that's far away and I understand we have to find a solution meanwhile.
> 
> What is the main problem you are trying to solve with this PR? Is it that you pass too many headers to downstream actions that don't need it therefore making them slower and use more bandwidth in remote execution? Is it about header visibility? Downstream rules not giving compilation error when using headers they shouldn't use?

All of these that you've indicated, plus one other weird one - it let's us manage command line include args, which we have seen exhaust kernel imposed limits on args count AND env var content sizes.
> 
> Also would this new attribute be used throughout your codebase or just in some key places? I ask this last question because we can do exactly the same thing your PR does already today in Bazel without any additional changes with the help of the C++ sandwich. If a lot of targets will be using the compile_only_deps attribute then the C++ sandwich idea would become really annoying.

We use it everywhere (>100s of instances) - we've had this feature available in bazels for 2+ years internally, and it would definitely be annoying to inject on top of our existing solutions
> 
> To summarize, the idea would be to have one cc_library with only the headers that you want to propagate. You would also have your cc_library with cc files depending on the previous headers library.
> 
> Now you introduce a new rule written in Starlark called cc_linking_library that would depend on just a single cc_library (wrapping the cc files library above) and omit the compilation context when creating a new [CcInfo](https://docs.bazel.build/versions/master/skylark/lib/CcInfo.html) 

I can see how all of this would work, I'm just not anxious to retool this after dragging it through 18 bazel versions.

**Comment by hlopko** [comment]
Pedro, which of these options do you prefer:

* accept the pr, + make it work with include scanning and extend the sandwich
* accept the pr guarded by an experimental flag + not extend the sandwich
* not accept the pr

Let's not leave the PR hanging in nowhere.

**Comment by werkt** [comment]
Ping. Can we come to some consensus on this?

File: src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCcLibraryRule.java [hunk 1]
```diff
@@ -33,6 +35,17 @@ public RuleClass build(RuleClass.Builder builder, RuleDefinitionEnvironment env)
...
         .requiresConfigurationFragments(CppConfiguration.class)
+        /*<!-- #BLAZE_RULE($cc_rule).ATTRIBUTE(compile_only_deps) -->
+        The list of other libraries to be linked in to the binary target.
+        This allows control over the requirement set as to whether a dependency
+        should contribute to the interface or only compilations of a library
+        <p>These can be <code>cc_library</code>, <code>cc_inc_library</code>, or
+        <code>objc_library</code> targets.</p>
+        <!-- #END_BLAZE_RULE.ATTRIBUTE -->*/
+        .add(attr("compile_only_deps", LABEL_LIST)
+            .allowedRuleClasses(BazelCppRuleClasses.DEPS_ALLOWED_RULES)
```
Comment (oquenchil) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCcLibraryRule.java]: Please change this to allow any rule with the provider CcInfo. Do not allow linker scripts.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java [hunk 2]
```diff
@@ -343,6 +344,14 @@ public CcCompilationHelper fromCommon(CcCommon common, ImmutableList<String> add
     return this;
   }
 
+  /** Sets fields that only exist in cc_library rules. */
+  public CcCompilationHelper fromLibrary(RuleContext ruleContext) {
```
Comment (oquenchil) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcCompilationHelper.java]: For public methods of CcCompilationHelper, please do not pass RuleContext attributes. The public methods of CcCompilationHelper are essentially the API of cc_common.compile() and we cannot accept RuleContext there. Just leave the public method here that takes a list of CcCompilationContexts.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java [hunk 3]
```diff
@@ -156,6 +156,7 @@ public static void init(
                 ccToolchain,
                 fdoContext)
             .fromCommon(common, additionalCopts)
+            .fromLibrary(ruleContext)
```
Comment (oquenchil) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java]: Use addImplementationCcCompilationContexts please.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java [hunk 4]
```diff
@@ -182,6 +183,7 @@ public static void init(
                 ruleContext.getFragment(CppConfiguration.class),
                 ruleContext.getSymbolGenerator())
             .fromCommon(ruleContext, common)
+            .fromLibrary(ruleContext)
```
Comment (oquenchil) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java]: Use addImplementationCcCompilationContexts please.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CcLinkingHelper.java [hunk 5]
```diff
@@ -178,6 +178,14 @@ public CcLinkingHelper fromCommon(RuleContext ruleContext, CcCommon common) {
     return this;
   }
 
+  /** Sets fields that only exist in cc_library rules. */
+  public CcLinkingHelper fromLibrary(RuleContext ruleContext) {
```
Comment (oquenchil) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/CcLinkingHelper.java]: Please remove this method and just pass every CcLinkingContext from CcLibrary.java as usual.

**Comment by werkt** [comment]
Another ping here.

**Comment by oquenchil** [comment]
This looks good to me. Marcel has some comments.

**Comment by r4nt** [comment]
I have some concerns on landing a change that so fundamentally changes how we provide C++ deps without a design doc review.
1. we do have much more than normal deps & compile-only deps; if we want to add those in the future, we should make sure we're not designing ourselves into a one-way street
2. high-level we need to decide whether this interacts with modules in a good way; I don't understand that from the description; comment in the code "The list of other libraries to be linked in to the binary target." makes this sound more like link-time deps?

**Comment by werkt** [comment]
> I have some concerns on landing a change that so fundamentally changes how we provide C++ deps without a design doc review.

This is discussed anecdotally in https://docs.google.com/document/d/1-tv0_79zGyBoDmaP_pYWaBVUwHUteLpAs90_rUl-VY8/edit#heading=h.ugb1hhxhyky4 and referred to in https://docs.google.com/document/d/1-tv0_79zGyBoDmaP_pYWaBVUwHUteLpAs90_rUl-VY8/edit#heading=h.ugb1hhxhyky4 under the Future Work section regarding header export.

> 
>     1. we do have much more than normal deps & compile-only deps; if we want to add those in the future, we should make sure we're not designing ourselves into a one-way street

The future has not yet come to pass here, and there are no controls for this. As discussed earlier in this thread, introducing this feature should not be prevented by, nor prevent a new mechanism from supplanting it, nor additional mechanisms being introduced to cover those cases.

> 
>     2. high-level we need to decide whether this interacts with modules in a good way; I don't understand that from the description; comment in the code "The list of other libraries to be linked in to the binary target." makes this sound more like link-time deps?

The transitive effect is that these libraries are "link-time deps", only affecting transitive consumer link actions. The compile action effect is limited to *only the library in which they are declared*, allowing the transitive downflow of headers (input files and include chains) from that dependency to be halted.

**Comment by r4nt** [comment]
> This is discussed anecdotally in https://docs.google.com/document/d/1-tv0_79zGyBoDmaP_pYWaBVUwHUteLpAs90_rUl-VY8/edit#heading=h.ugb1hhxhyky4 and referred to in https://docs.google.com/document/d/1-tv0_79zGyBoDmaP_pYWaBVUwHUteLpAs90_rUl-VY8/edit#heading=h.ugb1hhxhyky4 under the Future Work section regarding header export.

I'm aware of some of that history, but it doesn't look at the effect of this change in isolation.

> > ```
> > 1. we do have much more than normal deps & compile-only deps; if we want to add those in the future, we should make sure we're not designing ourselves into a one-way street
> > ```
> 
> The future has not yet come to pass here, and there are no controls for this. As discussed earlier in this thread, introducing this feature should not be prevented by, nor prevent a new mechanism from supplanting it, nor additional mechanisms being introduced to cover those cases.

Adding this feature will make it harder to change the meaning of the attribute name, and be a distraction all around if we go for a different route later (we'll need to spend time deprecating it and getting rid of it).
All code comes at a cost, especially in a system with many users. We've put more care into things that have significantly less influence of how folks describe deps at a fundamental level.

> > ```
> > 2. high-level we need to decide whether this interacts with modules in a good way; I don't understand that from the description; comment in the code "The list of other libraries to be linked in to the binary target." makes this sound more like link-time deps?
> > ```
> 
> The transitive effect is that these libraries are "link-time deps", only affecting transitive consumer link actions. The compile action effect is limited to _only the library in which they are declared_, allowing the transitive downflow of headers (input files and include chains) from that dependency to be halted. I don't know what 'modules' you refer to, but if it isn't tested, then by all means insist on a test. It certainly does not change how any library behaves that does not use the attribute.

To me "link-time deps" are only available at link time, but these are avaialble at compile time (of stuff in srcs I assume?), right?

**Comment by werkt** [comment]
> I'm aware of some of that history, but it doesn't look at the effect of this change in isolation.
> 

Forgot this one, which was mentioned in https://github.com/bazelbuild/bazel/issues/7310, where the behavior itself is described.
https://docs.google.com/document/d/1WW4iY3-74DVSX9wD3TiznrUIs7vybrxkj1U22JmWIrI/edit#heading=h.pjw8vw3mtrvy

The nominal effect is a noop without use. The effect with use is exactly as described and tested - compile actions derived from the the declaring cc_library rule see the transitive headers as inputs, consumers do not. Consumers see the link dependencies that they inject.

> Adding this feature will make it harder to change the meaning of the attribute name, and be a distraction all around if we go for a different route later (we'll need to spend time deprecating it and getting rid of it).

Nothing else does this now. Nor for the last three years that I've been asking for it. Deprecation is clearly not impossible, since it comes on every single release of bazel, and I doubt that anyone who decides to remove this working feature to be replaced by another will be incapable of handling such effort.

> All code comes at a cost, especially in a system with many users. We've put more care into things that have significantly less influence of how folks describe deps at a fundamental level.
> 

What is the quantitative measurement for 'more care' here? It would seem that 'many users', including myself, would set such a bar.

> To me "link-time deps" are only available at link time, but these are avaialble at compile time (of stuff in srcs I assume?), right?

The phrase `The list of other libraries to be linked in to the binary target.` was copied from the `deps` attribute of `[BazelCppRuleClasses.java](https://github.com/werkt/bazel/blob/2993bfbb8526c2bc8b75d2c200aab67a22d77b52/src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppRuleClasses.java#L271)`, where the `deps` list also means that it is available to members of `srcs`, in addition to tacitly becoming a part of the library's interface for consumers. The line was copied due to the "link semantics are unchanged" manner of the attribute's affect. The following lines expound on the ability to control header/include transitivity.

**Comment by r4nt** [comment]
> > I'm aware of some of that history, but it doesn't look at the effect of this change in isolation.
> 
> Forgot this one, which was mentioned in #7310, where the behavior itself is described.
> https://docs.google.com/document/d/1WW4iY3-74DVSX9wD3TiznrUIs7vybrxkj1U22JmWIrI/edit#heading=h.pjw8vw3mtrvy
> 
> The nominal effect is a noop without use. The effect with use is exactly as described and tested - compile actions derived from the the declaring cc_library rule see the transitive headers as inputs, consumers do not. Consumers see the link dependencies that they inject.

I don't think the description is helpful enough yet, and perhaps the name also could be better; an alternative might be "implemetation_only_deps" or "src_only_deps", as we do compile the header module / hdrs in some cases, and I we don't want them in there?

> > Adding this feature will make it harder to change the meaning of the attribute name, and be a distraction all around if we go for a different route later (we'll need to spend time deprecating it and getting rid of it).
> 
> Nothing else does this now. Nor for the last three years that I've been asking for it. Deprecation is clearly not impossible, since it comes on every single release of bazel, and I doubt that anyone who decides to remove this working feature to be replaced by another will be incapable of handling such effort.

I don't disagree with the need to have this, but the reason we don't have it yet is that nobody has been able to really drive consensus on this and offer a good strategy.

I would like to emphasize again that this is fundamentally changing how we provide & think about deps in bazel, and automation to manage build files will need to deal with this, if provided.
 
> > All code comes at a cost, especially in a system with many users. We've put more care into things that have significantly less influence of how folks describe deps at a fundamental level.
> 
> What is the quantitative measurement for 'more care' here? It would seem that 'many users', including myself, would set such a bar.

We do need to trade off users against ongoing maintenance cost, support story for that cost, etc. I agree that the change itself is small, but I think the surface of the change to the bazel user interace (build files) is quite large.
Usually in bazel we do design proposals that show the proposal and some rationalization why we don't want the alternatives, and capture why we think more comprehensive versions for example are currently not going to happen.

> > To me "link-time deps" are only available at link time, but these are avaialble at compile time (of stuff in srcs I assume?), right?
> 
> The phrase `The list of other libraries to be linked in to the binary target.` was copied from the `deps` attribute of `[BazelCppRuleClasses.java](https://github.com/werkt/bazel/blob/2993bfbb8526c2bc8b75d2c200aab67a22d77b52/src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppRuleClasses.java#L271)`, where the `deps` list also means that it is available to members of `srcs`, in addition to tacitly becoming a part of the library's interface for consumers. The line was copied due to the "link semantics are unchanged" manner of the attribute's affect. The following lines expound on the ability to control header/include transitivity.

Which brings me to the point that I think we need to adapt the docs for the normal deps, too.
The doc change says *what* you can do with compile_only_deps, but not *how*.

**Comment by werkt** [comment]
@r4nt so in the interest of building consensus and avoiding unnecessary work, I'll try to summarize.

- You feel the existing discussion is not sufficient nor justification and you will not approve (or will strike down) this change without a separate design document specifically for it (whether it be `compile_only_deps` or another name).
- This design doc needs to present a need, the alternatives, strategy, and must obtain consensus on the behavior of this change.
- The documentation for both `deps` and this both need to incorporate in the starlark comment instrumentation exactly what these different `deps`-like edges connote in the tree, and how to use them to effect control over header input (and other transitivity).

**Comment by r4nt** [comment]
@werkt I think that's a fair summary of my stance

**Comment by emusand** [comment]
We are interested in `compile_only_deps` in my organization as well. If `cc_library` A needs a dependency to compile its .c files, we do not want this dependency to leak to other libraries that depends on A.

I agree that include scanning will solve the build performance aspect, but users might still be able to include .h files in their source files that they should not have access to.

What is the way forward here? Is it anything that we in our organization can do to help?

**Comment by meisterT** [comment]
What's the status of this PR?

**Comment by oquenchil** [comment]
I brought this up with @lberki and I can write the design doc. I have never liked include scanning. Specifically, how deep ingrained include scanning is into the core of Bazel and how hard it is to pull the logic out to the leaves (the C++ rules implementation).

This change might not be enough to remove include scanning but perhaps it gets us closer. Even if we can't ever remove it, I believe there would still be performance benefits with this change internally at Google. Fewer headers passed along, less memory and less work for grep_includes?

It will be a while before I can merge this though if ever.

**Comment by oquenchil** [comment]
Closing this as it doesn't look likely that we will write the design doc or will work on this any time soon. I filed #12350 instead.

**Comment by catskul** [comment]
> Closing this as it doesn't look likely that we will write the design doc or will work on this any time soon. I filed #12350 instead.

Sad to see this. IMO this is a sorely needed feature. I'd vote to reopen if that was an option.

**Comment by catskul** [comment]
Oh, looks like something similar might have already landed in 6.2 as an experimental feature!:

https://bazel.build/reference/be/c-cpp#cc_library.implementation_deps

**Pull request closed without merging**